receiver/hostmetricsreceiver/internal/scraper/memoryscraper/memory_scraper.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memoryscraper
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/shirou/gopsutil/mem"
+	"go.opencensus.io/trace"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/pdata"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/pdatautil"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector/receiver/hostmetricsreceiver/internal"
+)
+
+// Scraper for Memory Metrics
+type Scraper struct {
+	config   *Config
+	consumer consumer.MetricsConsumer
+	cancel   context.CancelFunc
+}
+
+// NewMemoryScraper creates a set of Memory related metrics
+func NewMemoryScraper(ctx context.Context, cfg *Config, consumer consumer.MetricsConsumer) (*Scraper, error) {
+	return &Scraper{config: cfg, consumer: consumer}, nil
+}
+
+// Start
+func (c *Scraper) Start(ctx context.Context) error {
+	ctx, c.cancel = context.WithCancel(ctx)
+
+	go func() {
+		ticker := time.NewTicker(c.config.CollectionInterval())
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				c.scrapeMetrics(ctx)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Close
+func (c *Scraper) Close(ctx context.Context) error {
+	c.cancel()
+	return nil
+}
+
+func (c *Scraper) scrapeMetrics(ctx context.Context) {
+	ctx, span := trace.StartSpan(ctx, "memoryscraper.scrapeMetrics")
+	defer span.End()
```

Comment:
I was just looking around the repo a bit - should I be using `StartMetricsReceiveOp/EndMetricsReceiveOp` here instead of a custom trace?

Reply:
Would be better I think.

Reply:
Had a bit of a deeper look at this. From what I understand, `obsreport` is a helper package for generating standard traces & metrics for collector components, and currently uses the OpenCensus libraries for this.

A couple of questions if I want to use this here:

- I'm not sure what to provide for the "transport" attribute value (third parameter of `StartMetricsReceiveOp`). These metrics are scraped from the host rather than being "received" as such so it doesn't seem applicable - is it okay to pass an empty string?
- `obsreport.CountMetricPoints` is only implemented for the OC proto format. It probably needs a second implementation for the in-memory data structure rather than doing a transformation in the code (should be relatively straightforward to do, but probably better done in a separate PR)